cmd/clusterctl/pkg/client/cluster/template.go

```diff
@@ -0,0 +1,207 @@
+)
+
+// TemplateClient has methods to work with templates stored in the cluster/out of the provider repository.
+type TemplateClient interface {
+	// Returns a workload cluster template stored in the given ConfigMap.
+	GetFromConfigMap(configMapNamespace, configMapName, configMapDataKey, targetNamespace string, listVariablesOnly bool) (repository.Template, error)
+
+	// GetFromURL return a workload cluster reading from the given URL.
+	GetFromURL(templateURL, targetNamespace string, listVariablesOnly bool) (repository.Template, error)
+}
+
+// templateClient implements TemplateClient.
+type templateClient struct {
+	proxy               Proxy
+	configClient        config.Client
+	gitHubClientFactory func(configVariablesClient config.VariablesClient) (*github.Client, error)
+}
+
+// ensure templateClient implements TemplateClient.
+var _ TemplateClient = &templateClient{}
+
+// newTemplateClient returns a templateClient.
+func newTemplateClient(proxy Proxy, configClient config.Client) *templateClient {
+	return &templateClient{
+		proxy:               proxy,
+		configClient:        configClient,
+		gitHubClientFactory: getGitHubClient,
+	}
+}
+
+func (t *templateClient) GetFromConfigMap(configMapNamespace, configMapName, configMapDataKey, targetNamespace string, listVariablesOnly bool) (repository.Template, error) {
+	if configMapNamespace == "" {
+		return nil, errors.New("invalid GetFromConfigMap operation: missing configMapNamespace value")
+	}
+	if configMapName == "" {
+		return nil, errors.New("invalid GetFromConfigMap operation: missing configMapName value")
+	}
+
+	c, err := t.proxy.NewClient()
+	if err != nil {
+		return nil, err
+	}
+
+	configMap := &corev1.ConfigMap{}
+	configMapObjectKey := client.ObjectKey{
+		Namespace: configMapNamespace,
+		Name:      configMapName,
+	}
+
+	if err := c.Get(ctx, configMapObjectKey, configMap); err != nil {
+		return nil, errors.Wrapf(err, "error reading ConfigMap %s/%s", configMapNamespace, configMapName)
+	}
+
+	data, ok := configMap.Data[configMapDataKey]
+	if !ok {
+		return nil, errors.Errorf("the ConfigMap %s/%s does not have the %s data key", configMapNamespace, configMapName, configMapDataKey)
+	}
+
+	return repository.NewTemplate([]byte(data), t.configClient.Variables(), targetNamespace, listVariablesOnly)
+}
+
+func (t *templateClient) GetFromURL(templateURL, targetNamespace string, listVariablesOnly bool) (repository.Template, error) {
+	if templateURL == "" {
+		return nil, errors.New("invalid GetFromURL operation: missing templateURL value")
+	}
+
+	content, err := t.getURLContent(templateURL)
+	if err != nil {
+		return nil, err
+	}
+
+	return repository.NewTemplate(content, t.configClient.Variables(), targetNamespace, listVariablesOnly)
+}
+
+func (t *templateClient) getURLContent(templateURL string) ([]byte, error) {
+	rURL, err := url.Parse(templateURL)
+	if err != nil {
+		return nil, errors.Wrapf(err, "invalid GetFromURL operation: failed to parse %q", templateURL)
```

Comment:
```suggestion
		return nil, errors.Wrapf(err, "failed to parse %q", templateURL)
```
We can then wrap this error on line 104 with `errors.Wrapf(err, "invalid GetFromURL operation")`. 
This way the other errors that may be returned below are also wrapped with `invalid GetFromURL operation`.